tests/operator_test.go

```diff
@@ -206,4 +209,91 @@ var _ = Describe("Operator delete CDI tests", func() {
 		err = f.CdiClient.CdiV1alpha1().CDIs().Delete(cr.Name, &metav1.DeleteOptions{DryRun: []string{"All"}})
 		Expect(err).ToNot(HaveOccurred())
 	})
+
+	It("should deploy components that tolerate CriticalAddonsOnly taint", func() {
+		var cdiPods *corev1.PodList
+		var err error
+		By("finding all nodes that are running cdi components", func() {
+			cdiPods, err = f.K8sClient.CoreV1().
+				Pods(f.CdiInstallNs).List(metav1.ListOptions{})
+			Expect(err).ShouldNot(HaveOccurred(), "failed listing cdi pods")
+			Expect(len(cdiPods.Items)).
+				To(BeNumerically(">", 0), "no cdi pods found")
+		})
+
+		// nodes that run cdi components.
+		cdiNodes := make(map[string]*corev1.Node)
+		By("adding taints to any node that runs cdi component", func() {
+			criticalPodTaint := corev1.Taint{
+				Key:    "CriticalAddonsOnly",
+				Value:  "",
+				Effect: corev1.TaintEffectNoExecute,
+			}
+
+			for _, cdiPod := range cdiPods.Items {
+				cdiNode, err := f.K8sClient.CoreV1().Nodes().Get(cdiPod.Spec.NodeName, metav1.GetOptions{})
+				Expect(err).ShouldNot(HaveOccurred(), "failed retrieving node")
+				hasTaint := false
+				// check if node already has the taint
+				for _, taint := range cdiNode.Spec.Taints {
+					if reflect.DeepEqual(taint, criticalPodTaint) {
+						// node already have the taint set
+						hasTaint = true
+						break
+					}
+				}
+				if hasTaint {
+					continue
+				}
+				cdiNode.ResourceVersion = ""
+				cdiNodes[cdiPod.Spec.NodeName] = cdiNode
+				cdiNodeCopy := cdiNode.DeepCopy()
+				cdiNodeCopy.Spec.Taints = append(cdiNodeCopy.Spec.Taints, criticalPodTaint)
+				_, err = f.K8sClient.CoreV1().Nodes().Update(cdiNodeCopy)
+				Expect(err).ShouldNot(HaveOccurred(), "failed setting taint on node")
+			}
+		})
+
+		defer func() {
+			var errors []error
+			By("restoring nodes")
+			for _, nodeSpec := range cdiNodes {
+				_, err := f.K8sClient.CoreV1().Nodes().Update(nodeSpec)
```

Comment:
Wouldn't it be better to just remove the taint instead of overwriting the whole spec?  Could there be unwanted/unexpected changes as a result of this approach?

Reply:
Theoretically, yes, in a multi-user/multi-threaded environment, we could expose ourselves to a race. But in this case, it's a functional test that can't run in a multi-threaded environment anyway so I don't see a risk here.

If you think that patching is better here, I don't mind to use a patch instead of an update.

Reply:
I think update is fine.